plugins/algoltek-aux/fu-algoltek-aux-device.c

```diff
@@ -0,0 +1,560 @@
+	guint32 firmwareBufferCount = 0;
+	guint16 WCRC = 0x1021;
+
+	g_autoptr(GInputStream) stream_isp = NULL;
+	g_autoptr(GInputStream) stream_payload = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 18, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 2, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 80, NULL);
+
+	if (!fu_algoltek_aux_device_en(self, error))
+		return FALSE;
+	
+	if (!fu_algoltek_aux_device_rst(self, error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 500);
+
+	if (!fu_algoltek_aux_device_wrr(self, 0x80AD, 0, error))
+		return FALSE;
+
+	if (!fu_algoltek_aux_device_wrr(self, 0x80C0, 0, error))
+		return FALSE;
+
+	if (!fu_algoltek_aux_device_wrr(self, 0x80C9, 0, error))
+		return FALSE;
+
+	if (!fu_algoltek_aux_device_wrr(self, 0x80D1, 0, error))
+		return FALSE;
+
+	if (!fu_algoltek_aux_device_wrr(self, 0x80D9, 0, error))
+		return FALSE;
+
+	if (!fu_algoltek_aux_device_wrr(self, 0x80E1, 0, error))
+		return FALSE;
+
+	if (!fu_algoltek_aux_device_wrr(self, 0x80E9, 0, error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 20);
+
+	// /* get ISP image */
+	stream_isp = fu_firmware_get_image_by_id_stream(firmware, "isp", error);
+	if (stream_isp == NULL)
+		return FALSE;
+	
+    if (!fu_algoltek_aux_device_isp(self, stream_isp, &WCRC, fu_progress_get_child(progress), error))
+				return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 2000);
+#if 1
+	if (!fu_algoltek_aux_device_bot(self, 0x6000, error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 2000);
+
+	if (!fu_algoltek_aux_device_ers(self, error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 5000);
+
+	/* get payload image */
+	stream_payload = fu_firmware_get_image_by_id_stream(firmware, FU_FIRMWARE_ID_PAYLOAD, error);
+	if (stream_payload == NULL)
+		return FALSE;
+
+
+#if 1
+    if (!fu_algoltek_aux_device_wrf(self, stream_payload, &WCRC, fu_progress_get_child(progress), error))
+	    return FALSE;
+
+#endif
+	
+#endif
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_aux_device_setup(FuDevice *device, FuFirmware *firmware, GError **error)
+{
+	FuAlgoltekAuxDevice *self = FU_ALGOLTEK_AUX_DEVICE(device);
+	g_autofree gchar *version_str = NULL;
+	g_autoptr(GByteArray) version_data = NULL;
+	g_autoptr(GByteArray) guid_data = g_byte_array_new();
+	fu_byte_array_set_size(guid_data, 4, 0x0);
+
+	/* AuxDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_algoltek_aux_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	version_data = fu_algoltek_aux_device_rdv(self, error);
+
+	g_print("\nSETUP\n");
+	if (version_data == NULL) {
+		return NULL;
+	}
+
+	version_str = g_strdup_printf("%s", version_data->data);
+
+	fu_device_set_version(FU_DEVICE(self), version_str);
+	g_autofree gchar *product = NULL;
+	
+	product = g_strdup_printf("MST-ALGOLTEK");
+	
+	
+	fu_device_add_instance_id(FU_DEVICE(self), product);
```

Comment:
this isn't really a product string -- you need to get something from the hardware.

Reply:
Got it.
I have currently modified it to read information from the hardware.